week3/tideman/tideman.c

```diff
@@ -0,0 +1,268 @@
+void record_preferences(int ranks[]);
+void add_pairs(void);
+void sort_pairs(void);
+void lock_pairs(void);
+void print_winner(void);
+int get_rank_index(int candidateNumber, int ranks[]);
+
+int main(int argc, string argv[]) {
+    bool thereIsNoCandidateInput = argc < 2;
+
+    if (thereIsNoCandidateInput) {
+        printf("Usage: tideman [candidate ...]\n");
+
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    bool candidateOverflow = candidate_count > MAX;
+
+    if (candidateOverflow) {
+        printf("Maximum number of candidates is %i\n", MAX);
+
+        return 2;
+    }
+
+    for (int candidateIndex = 0; candidateIndex < candidate_count; candidateIndex++) {
+        candidates[candidateIndex] = argv[candidateIndex + 1];
+    }
+
+    for (int rowIterator = 0; rowIterator < candidate_count; rowIterator++) {
+        for (int columnIterator = 0; columnIterator < candidate_count; columnIterator++) {
+            locked[rowIterator][columnIterator] = false;
+        }
+    }
+
+    pair_count = 0;
+    int voter_count = get_int("Number of voters: ");
+
+    for (int voterIterator = 0; voterIterator < voter_count; voterIterator++) {
+        int ranks[candidate_count];
+
+        for (int candidateRank = 0; candidateRank < candidate_count; candidateRank++) {
+            string name = get_string("Rank %i: ", candidateRank + 1);
+            bool voteIsInvalid = !vote(candidateRank, name, ranks);
+
+            if (voteIsInvalid) {
+                printf("Invalid vote.\n");
+
+                return 3;
+            }
+        }
+       
+        record_preferences(ranks);
+        printf("\n");
+    }
+
+    add_pairs();
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+
+    return 0;
+}
+
+bool vote(int candidateRank, string name, int ranks[]) {
+    for (int i = 0; i < candidate_count; i++) {
+        bool candidateIsValid = strcmp(candidates[i], name) == 0;
+
+        if (candidateIsValid) {
+            ranks[candidateRank] = i;
+
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void record_preferences(int ranks[]) {
+    for (int rowIterator = 0; rowIterator < candidate_count; rowIterator++) {
+        for (int columnIterator = 0; columnIterator < candidate_count; columnIterator++) {
+            int firstCompareCandidate = rowIterator;
+            int secondCompareCandidate = columnIterator;
+            int indexOfFirstCompareCandidate = get_rank_index(firstCompareCandidate, ranks);
+            int indexOfSeconCompareCandidate = get_rank_index(secondCompareCandidate, ranks);
+            bool candidateIsPreferred = indexOfFirstCompareCandidate < indexOfSeconCompareCandidate;
+
+            if (candidateIsPreferred) {
+                preferences[rowIterator][columnIterator]++;
+            }
+        }
+    }
+
+    return;
+}
+
+void add_pairs(void) {
+    int jparser = 1;
+
+    for (int i = 0; i < candidate_count; i++) {
+        for (int j = jparser; j < candidate_count; j++) {
```

Comment:
```suggestion
        for (int j = i +1; j < candidate_count; j++) {
```
This makes it clearer that you always want j to be at least 1 higher than i. That way you dont need jparser.